From f75f66e6f992e28fed5e4ed4e799e3f3a243b3fb Mon Sep 17 00:00:00 2001
From: Roshan Pius <rpius@google.com>
Date: Thu, 29 Jan 2026 14:22:02 -0800
Subject: [PATCH] nfc: Unbind from previously set foreground pref service

This was a bug introduced via b/365533082.
If an app with HCE service registered invokes "setPrefferedService" and
then moves to the background / closed, then NFC stack keeps the
foreground service binding alive until the next NFC transaction (or
reboot).

Bug: 479865395
Bug: 473814675
Flag: EXEMPT BUGFIX
Test: Manual verification to ensure the service is unbound
when the app moves to the background
Test: atest CtsNfcHceMultiDeviceTestCases  -- --testparam
pn532_serial_path=/dev/ttyUSB0

Change-Id: Ifca13644d6071b6bbf7fb91122724c9dfa42bf27
Merged-In: Ifca13644d6071b6bbf7fb91122724c9dfa42bf27
(cherry picked from commit 2711c9ef108434aa87e92a3de1e1b439f8d07c4f)
---
 .../cardemulation/HostEmulationManager.java   | 81 ++++++++++++-------
 1 file changed, 54 insertions(+), 27 deletions(-)

diff --git a/NfcNci/src/com/android/nfc/cardemulation/HostEmulationManager.java b/NfcNci/src/com/android/nfc/cardemulation/HostEmulationManager.java
index 25c55fd00..866706163 100644
--- a/NfcNci/src/com/android/nfc/cardemulation/HostEmulationManager.java
+++ b/NfcNci/src/com/android/nfc/cardemulation/HostEmulationManager.java
@@ -87,6 +87,9 @@ import java.util.Random;
 import java.util.Set;
 import java.util.regex.Pattern;
 
+/**
+ * TODO(b/479895625): Remove the {@link Flags#allowMultipleHceBindings()} check from the codebase.
+ */
 public class HostEmulationManager {
     static final String TAG = "HostEmulationManager";
     static final boolean DBG = NfcProperties.debug_enabled().orElse(true);
@@ -247,6 +250,35 @@ public class HostEmulationManager {
         }
     };
 
+    private void unbindInactiveServicesLocked() {
+        ComponentNameAndUser preferredNameAndUser = mAidCache.getPreferredService();
+        Map<ComponentNameAndUser, HostEmulationConnection> retainedConnections =
+                new HashMap<>();
+        mComponentNameToConnectionsMap.keySet().forEach((key) -> {
+            if (!preferredNameAndUser.equals(key)) {
+                HostEmulationConnection connection =
+                        mComponentNameToConnectionsMap.get(key);
+                if (connection.mMessenger != null) {
+                    Log.d(TAG, "unbindServiceInactiveServicesLocked: service "
+                            + connection.mServiceConnection);
+                    try {
+                        mContext.unbindService(connection.mServiceConnection);
+                    } catch (IllegalArgumentException iae) {
+                        Log.wtf(TAG,
+                                "unbindInactiveServicesLocked: "
+                                        + "Exception while unbinding "
+                                        + key.getComponentName()
+                                        + " service connection",
+                                iae);
+                    }
+                }
+            } else {
+                retainedConnections.put(key, mComponentNameToConnectionsMap.get(key));
+            }
+        });
+        mComponentNameToConnectionsMap = retainedConnections;
+    }
+
     Runnable mUnbindInactiveServicesRunnable =
             new Runnable() {
                 @Override
@@ -261,32 +293,6 @@ public class HostEmulationManager {
                     }
                 }
 
-                void unbindInactiveServicesLocked() {
-                    ComponentNameAndUser preferredNameAndUser = mAidCache.getPreferredService();
-                    Map<ComponentNameAndUser, HostEmulationConnection> retainedConnections =
-                            new HashMap<>();
-                    mComponentNameToConnectionsMap.keySet().forEach((key) -> {
-                        if (!preferredNameAndUser.equals(key)) {
-                            HostEmulationConnection connection =
-                                mComponentNameToConnectionsMap.get(key);
-                            if (connection.mMessenger != null) {
-                                try {
-                                    mContext.unbindService(connection.mServiceConnection);
-                                } catch (IllegalArgumentException iae) {
-                                    Log.wtf(TAG,
-                                            "unbindInactiveServicesLocked: "
-                                                    + "Exception while unbinding "
-                                                    + key.getComponentName()
-                                                    + " service connection",
-                                            iae);
-                                }
-                            }
-                        } else {
-                            retainedConnections.put(key, mComponentNameToConnectionsMap.get(key));
-                        }
-                    });
-                    mComponentNameToConnectionsMap = retainedConnections;
-                }
             };
 
     // Runnable to re-enable observe mode after a transaction. This should be delayed after
@@ -766,7 +772,7 @@ public class HostEmulationManager {
             if (service != null) {
                 bindServiceIfNeededLocked(userId, service);
             } else {
-                unbindServiceIfNeededLocked();
+                unbindServiceIfNeededLocked(/* force */ true);
             }
          }
      }
@@ -1370,8 +1376,29 @@ public class HostEmulationManager {
         }
     }
 
+    /**
+     * Regular unbind to be used when delivering polling loops, etc and switching between services
+     * during a single transaction. This retains the binding to avoid trashing while the transaction
+     * is in progress. The bindings are cleared in that case when
+     * {@link #mUnbindInactiveServicesRunnable} runs after the completion of transaction.
+     */
     void unbindServiceIfNeededLocked() {
+        unbindServiceIfNeededLocked(/* force */ false);
+    }
+
+    /**
+     * Unbind regular app service.
+     *
+     * @param force Force unbind is used for cases where the preferred service is unset.
+     * Otherwise, we are at the mercy of the next run of {@link #mUnbindInactiveServicesRunnable}
+     * invocation which is only triggered when the next NFC transaction occurs.
+     */
+    void unbindServiceIfNeededLocked(boolean force) {
         if (isMultipleBindingSupported()) {
+            if (force) {
+                unbindInactiveServicesLocked();
+                return;
+            }
             if (mServiceName == null
                     || CompatChanges.isChangeEnabled(
                         DONT_IMMEDIATELY_UNBIND_SERVICES,
-- 
2.55.0.897.gb25b4bd76c-goog

